packages/liveblocks-react-ui/src/primitives/Composer/index.tsx

```diff
@@ -1001,13 +1028,45 @@ const ComposerForm = forwardRef<HTMLFormElement, ComposerFormProps>(
     forwardedRef
   ) => {
     const Component = asChild ? Slot : "form";
-    const room = useRoom();
     const [isEmpty, setEmpty] = useState(true);
     const [isSubmitting, setSubmitting] = useState(false);
     const [isFocused, setFocused] = useState(false);
+
+    const composerRoomId = useContext(ComposerRoomIdContext);
+    const thread = useContext(ThreadDataContext);
+    const comment = useContext(CommentDataContext);
+    const room = useRoomOrNull();
+
+    function getRoomId() {
+      // If the Compser.Form is within a Composer context, we simply return the same room id
+      if (composerRoomId) {
+        return composerRoomId;
+      }
+      // If Composer.Form is within a Comment context (editing a comment)
+      else if (comment) {
+        return comment.roomId;
+      }
+      // If Composer.Form is within a Thread context (replying to a thread)
+      else if (thread) {
+        return thread.roomId;
+      }
+      // If Composer.Form is within a Room context (creating a new thread)
+      else if (room) {
+        return room.id;
+      }
+
+      // If none of the contexts are available, throw an error
+      throw new Error(
+        "Composer.Form must be a descendant of RoomProvider, Thread, or Comment component."
+      );
+    }
```

Comment:
The process of obtaining room id for a `Composer.Form` is simpler than obtaining one for `Composer` because unlike `Composer`, `Composer.Form` doesn't accept props like `threadId` and `commentId`. So, we'd simply check if the `Composer.Form` is a descendant of `Composer`, in that case, we use the room id passed from `Composer`, otherwise, we check for the thread, comment and room context in order. If no context is available, we throw an error.

Reply:
This does make the primitives somewhat linked to the default components and less independent, could something like an `@internal` `roomId` prop on `Composer.Form` prop also work? Which would be populated by `Composer`.

Reply:
+1, I like this approach. Having the props passed in makes this implementation simpler and more direct. If they're not passed in, we can look it up from the `RoomProvider`. If not, it's an error.